README.md
[wiki-dtd]: https://en.wikipedia.org/wiki/Document_type_definition
[dtd-data]: ./src/data.ts

<div align="center">

# Email Doctype

Easily manage your email's [Document Type Definition (DTD)][wiki-dtd].

[![](https://img.shields.io/npm/v/email-doctype?color=black&logoColor=white)](https://www.npmjs.com/package/email-doctype)
![](https://img.shields.io/github/workflow/status/email-types/doctype/Test?logo=github&color=black)
[![](https://img.shields.io/codecov/c/github/email-types/doctype?logo=codecov&logoColor=white&color=black)](https://codecov.io/gh/email-types/doctype)
[![](https://img.shields.io/bundlephobia/min/email-doctype?logo=bundlephobia&color=black)](https://bundlephobia.com/result?p=email-doctype)
[![](https://img.shields.io/bundlephobia/minzip/email-doctype?logo=bundlephobia&color=black)](https://bundlephobia.com/result?p=email-doctype)

</div>

Install with yarn

```sh
yarn add email-doctype
```

Or install with npm:

```sh
npm install email-doctype
```

## Doctypes

| Keys                     | Alias Prefix   | Alias Version | Alias Suffix          |
| :----------------------- | :------------- | :------------ | :-------------------- |
| `html-5`                 | `h` or `html`  | `5`           | n/a                   |
| `html-4.01-strict`       | `h` or `html`  | `4.01`        | `s` or `strict`       |
| `html-4.01-transitional` | `h` or `html`  | n/a           | `t` or `transitional` |
| `xhtml-1.1`              | `x` or `xhtml` | `1.1`         | n/a                   |
| `xhtml-1.0-strict`       | `x` or `xhtml` | `1` or `1.0`  | `s` or `strict`       |
| `xhtml-1.0-transitional` | `x` or `xhtml` | n/a           | `t` or `transitional` |

**Aliases**

While you can easily get a doctype using a `Key`, you can also use an alias. An
alias can be an `Alias Version` or string composing of an `Alias Prefix`,
`Alias Version`, and `Alias Suffix`.

- Casing will be normalized so `keys` and `aliases` are case insensitive
- Whitespace is removed so spaces are allowed
- When there is an `Alias Suffix` available but not used, the returned doctype
  will use the `strict` alias suffix

**Examples**

Checkout the [test](./test/index.test.ts) for more examples.

```ts
import * as doctype from 'email-doctype';

doctype.version('recommended');
doctype.version('html-1.0-transitional');
doctype.version('XHTML 1.0 Transitional');
// => '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">'

doctype.version(5);
doctype.version('html-5');
doctype.version('HTML 5');
// => '<!DOCTYPE html>'

doctype.version(4.01);
doctype.version('xhtml-4.01-strict');
// => '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">'

doctype.version(1.1);
doctype.version('xhtml-1.1');
doctype.version('XHTML 1.1');
// => '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">'

doctype.version(1.0);
doctype.version('xhtml-1.0-strict');
doctype.version('XHTML 1.0 Strict');
// => '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">'
```

## Extracting a doctype

Extract the doctype from a given html string.

```ts
import { extract } from 'email-doctype';

const html = '<!DOCTYPE html> <html><head></head><body></body></html>';

extract(html);
// => '<!DOCTYPE html>'
```

## Setting a doctype

Add a doctype to a given html string.

```ts
import { set } from 'email-doctype';

const html = '<html><head></head><body></body></html>';

set(html, 'HTML 5');
// => '<!DOCTYPE html> <html><head></head><body></body></html>'
```

**Automatic replacement**

When the given html contains a doctype, it will be automatically replaced with
your new doctype.

```ts
import { set } from 'email-doctype';

const html = '<!DOCTYPE html> <html><head></head><body></body></html>';

set(html, 'HTML 4.01');
// '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd"> <html><head></head><body></body></html>'
```

## Removing a doctype

Remove a given html string's doctype, returning just the html.

```ts
import { remove } from 'email-doctype';

const html = '<!DOCTYPE html> <html><head></head><body></body></html>';

remove(html);
// => '<html><head></head><body></body></html>'
```

## API

### version

```ts
version(alias: string): string
```

Gets a doctype from a given alias.

### extract

```ts
extract(html: string): string
```

Extracts the doctype from a given html string.

### set

```ts
set(html: string, alias: string = 'recommended'): string
```

Adds a doctype to a given html string.

### remove

```ts
remove(html: string): string
```

Removes a given html string's doctype, returning just the html.


CHANGELOG.md
### 0.0.1-canary.2 - 2019-12-23

#### 📦 Dependencies

- Update all deps to latest (#3)
  ([7ff877a](https://github.com/email-types/doctype/commit/7ff877a)), closes
  [#3](https://github.com/email-types/doctype/issues/3)

### 0.0.1-canary.1 - 2019-12-16

#### 🐞 Fixes

- Readme link (#2)
  ([22e4ce8](https://github.com/email-types/doctype/commit/22e4ce8)), closes
  [#2](https://github.com/email-types/doctype/issues/2)

## 0.0.1-canary.0 - 2019-12-16

#### 🚀 Updates

- Initial doctypes
  ([8dbd1de](https://github.com/email-types/doctype/commit/8dbd1de))

#### 🐞 Fixes

- Point to dist folder (#1)
  ([2114b9c](https://github.com/email-types/doctype/commit/2114b9c)), closes
  [#1](https://github.com/email-types/doctype/issues/1)


test/set.test.ts
import { set, version } from '../src';

const html = '<html><head><title>Title</title></head><body>Hello</body></html>';

describe('doctype', () => {
  it('returns with the recommended doctype', () => {
    const result = set(html);
    expect(result).toBe(`${version('xhtml-1.0-transitional')} ${html}`);
  });

  it('returns with a doctype', () => {
    const result = set(html, 'HTML 5');
    expect(result).toBe(`${version('html-5')} ${html}`);
  });

  it('returns with a doctype using an alias', () => {
    const result = set(html, 5);
    expect(result).toBe(`${version('html-5')} ${html}`);
  });

  it('replaces current doctype with the recommended doctype', () => {
    const result = set(`<!DOCTYPE html> ${html}`);
    expect(result).toBe(`${version('xhtml-1.0-transitional')} ${html}`);
  });
});


test/index.test.ts
import * as doctype from '../src';

const specs: Array<{
  name: string;
  queries: doctype.Alias[];
  expected: string;
}> = [
  {
    name: 'returns the recommended doctype',
    queries: ['recommended'],
    expected: doctype.version('xhtml-1.0-transitional'),
  },
  {
    name: 'returns an HTML 5 doctype',
    queries: [5, '5', 'h 5', 'html 5', 'html-5', 'HTML 5'],
    expected: doctype.version('html-5'),
  },
  {
    name: 'returns an HTML 4.01 Strict doctype',
    queries: [
      4,
      4.01,
      '4.01',
      'h 4.01',
      'h-4.01',
      'html 4.01',
      'html-4.01',
      '4.01s',
      '4.01-s',
      'h 4.01 s',
      'h-4.01-s',
      'html 4.01 strict',
      'html-4.01-strict',
      'HTML 4.01 Strict',
    ],
    expected: doctype.version('html-4.01-strict'),
  },
  {
    name: 'returns an HTML 4.01 Transitional doctype',
    queries: [
      '4.01t',
      '4.01-t',
      'h 4.01 t',
      'h-4.01-t',
      'html 4.01 transitional',
      'html-4.01-transitional',
      'HTML 4.01 Transitional',
    ],
    expected: doctype.version('html-4.01-transitional'),
  },
  {
    name: 'returns an XHTML 1.1 doctype',
    queries: [
      1.1,
      1.1,
      '1.1',
      'x 1.1',
      'x-1.1',
      'xhtml 1.1',
      'xhtml-1.1',
      'XHTML 1.1',
    ],
    expected: doctype.version('xhtml-1.1'),
  },
  {
    name: 'returns an XHTML 1.0 Strict doctype',
    queries: [
      1,
      1.0,
      '1.0',
      'x 1.0',
      'x-1.0',
      'xhtml 1.0',
      'xhtml-1.0',
      '1.0s',
      '1.0-s',
      'x 1.0 s',
      'x-1.0-s',
      'xhtml 1.0 strict',
      'xhtml-1.0-strict',
      'XHTML 1.0 Strict',
    ],
    expected: doctype.version('xhtml-1.0-strict'),
  },
  {
    name: 'returns an XHTML 1.0 Transitional doctype',
    queries: [
      '1.0t',
      '1.0-t',
      'x 1.0 t',
      'x-1.0-t',
      'xhtml 1.0 transitional',
      'xhtml-1.0-transitional',
      'XHTML 1.0 Transitional',
    ],
    expected: doctype.version('xhtml-1.0-transitional'),
  },
];

describe('doctype', () => {
  specs.forEach((spec) => {
    it(spec.name, () => {
      spec.queries.forEach((query) => {
        const result = doctype.version(query);
        expect(result).toBe(spec.expected);
      });
    });
  });

  it('throws an error when given an incorrect name', () => {
    expect(() => doctype.version('foobar')).toThrow(
      '"foobar" is not a valid doctype name',
    );
  });
});


test/remove.test.ts
import { remove, version } from '../src';

const html = '<html><head><title>Title</title></head><body>Hello</body></html>';

describe('doctype set', () => {
  it('removes doctype from given html', () => {
    const result = remove(`${version('html-5')} ${html}`);
    expect(result).toBe(html);
  });

  it('returns html when there is not doctype to remove', () => {
    const result = remove(html);
    expect(result).toBe(html);
  });
});


.eslintrc.js
const extensions = ['.ts', '.tsx', '.js', '.jsx', '.json'];

module.exports = {
  extends: [
    '@postmates',
    'prettier',
    'prettier/standard',
    'prettier/@typescript-eslint',
  ],
  plugins: ['prettier'],
  env: { node: true },
  settings: {
    'import/extensions': extensions,
    'import/resolver': {
      node: { extensions },
    },
  },
  rules: {
    '@typescript-eslint/explicit-function-return-type': 0,
  },
  overrides: [
    {
      files: ['*.test.ts'],
      extends: ['plugin:jest/recommended', 'plugin:jest/style'],
      plugins: ['jest'],
      env: { jest: true },
    },
    {
      files: ['scripts/**/*'],
      rules: {
        'no-console': 'off',
        'import/no-extraneous-dependencies': [
          'error',
          {
            devDependencies: true,
            optionalDependencies: false,
          },
        ],
      },
    },
  ],
};


src/index.ts
import { Version, doctypes } from './data';

type GlobalsString = string & { _?: never };

type GlobalsNumber = number & { _?: never };

type Alias = Version | GlobalsString | GlobalsNumber;

const clean = (alias: Alias): string =>
  String(alias)
    .toLowerCase()
    .replace(/\.(01|0)/, '')
    .replace(/\b(html|transitional|xhtml)\b/g, (_: unknown, v: string) =>
      v.charAt(0),
    )
    .replace(/s(?:trict)?/, '')
    .replace(/(-|\.|\s)+/g, '');

const aliases = Object.keys(doctypes).reduce<Record<string, string>>(
  (acc, key) => {
    acc[clean(key)] = key;
    return acc;
  },
  {},
);

export const version = (alias: Alias): string => {
  const key = clean(alias);
  const name = aliases[key] || aliases[`h${key}`] || aliases[`x${key}`];
  const dtd = (doctypes as typeof aliases)[name] || doctypes[alias as Version];

  if (!dtd) {
    throw Error(`"${alias}" is not a valid doctype name`);
  }

  return dtd;
};

export const extract = (html: string): string | undefined => {
  const pattern = /<\s*!\s*doctype[^>]*>/im;
  const match = pattern.exec(html);
  return (match && match[0]) || undefined;
};

export const set = (html: string, alias: Alias = 'recommended'): string => {
  const curr = extract(html);
  const next = version(alias);
  return !curr ? `${next} ${html}` : html.replace(curr, next);
};

export const remove = (html: string): string => {
  const curr = extract(html);
  return !curr ? html : html.replace(curr, '').trim();
};

export { Version, Alias };


scripts/generate.ts
import { EOL } from 'os';
import { resolve } from 'path';
import { outputFile } from 'fs-extra';
import { format, resolveConfig } from 'prettier';
import { cyan, green, red } from 'colors';
import { doctypes as rawDoctypes, Doctype } from '@email-types/data/html';

const cwd = process.cwd();
const output = resolve(__dirname, '../src/data.ts');

const orderedKeys: (keyof Doctype)[] = [
  'element',
  'keyword',
  'publicIdentifier',
  'systemIdentifier',
];

const toKebabCase = (value: string): string => {
  return value.toLowerCase().replace(/\s/g, '-');
};

const coerce = (doctype: Doctype): string => {
  const dtd = orderedKeys.reduce((acc, key) => {
    const value = doctype[key];

    if (!value) return acc;

    if (key === 'publicIdentifier' || key === 'systemIdentifier') {
      return `${acc} "${value}"`;
    }

    return `${acc} ${value}`;
  }, '');

  return `<!DOCTYPE${dtd}>`;
};

const versions: string[] = [];
const doctypes = rawDoctypes.reduce<Record<string, string>>((acc, doctype) => {
  const key = toKebabCase(doctype.name);
  versions.push(`'${key}'`);
  acc[key] = coerce(doctype);
  return acc;
}, {});

doctypes.recommended = doctypes['xhtml-1.0-transitional'];
versions.push(`'recommended'`);

const data = [
  '/* This file is automatically generated and should not be edited directly. */',
  `export type Version = ${versions.join(' | ')};`,
  `export const doctypes: Record<Version, string> = ${JSON.stringify(
    doctypes,
  )};`,
].join(`${EOL}${EOL}`);

(async (): Promise<void> => {
  try {
    console.log(`${cyan('[ info ]')} generating data...`);
    const config = await resolveConfig(cwd);
    const content = format(data, {
      parser: 'typescript',
      ...config,
    });

    await outputFile(output, content.replace(/\n/g, EOL));
    console.log(`${green('[ done ]')} successfully generated data`);
  } catch (err) {
    console.log(`${red('[ error ]')} ${err}`);
    process.exit(0);
  }
})();


src/data.ts
/* This file is automatically generated and should not be edited directly. */

export type Version =
  | 'html-5'
  | 'html-4.01-strict'
  | 'html-4.01-transitional'
  | 'xhtml-1.1'
  | 'xhtml-1.0-strict'
  | 'xhtml-1.0-transitional'
  | 'recommended';

export const doctypes: Record<Version, string> = {
  'html-5': '<!DOCTYPE html>',
  'html-4.01-strict':
    '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN" "http://www.w3.org/TR/html4/strict.dtd">',
  'html-4.01-transitional':
    '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">',
  'xhtml-1.1':
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">',
  'xhtml-1.0-strict':
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">',
  'xhtml-1.0-transitional':
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">',
  recommended:
    '<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">',
};
